number of islands/Program.cs

```diff
@@ -0,0 +1,149 @@
+﻿
+/*
+Number of Islands
+
+Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.
+
+An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.
+
+1 1 0 0 0 
+1 1 0 0 0 
+0 0 1 0 0 
+0 0 0 1 1
+
+ex. 1. Number of islands: 3
+
+1 0 1 1 1 
+1 0 1 0 1 
+1 0 1 0 1 
+1 1 1 0 1
+
+ex. 2. Number of islands: 1
+
+Solution
+
+Matrix of ones and zeroes can be viewed as a graph,
+where each node (cell) is connected with one to four other nodes (cells). 
+That is, to the right, downwards, to the left, and upwards.
+
+We will iterate over matrix, looking for a "piece of land".
+
+When a land is found, we will "explore" the island by looking for adjacent land pieces with DFS (depth-first-search).
+
+DFS means, that we will put every adjacent piece of land to the stack.
+We will then pop each piece from stack for processing, and eventually, we will visit them all.
+To mind visited indices, we will put coordinate tuples of (x, y) into a "visited" dictionary.
+
+So, when we found an "unvisited" island:
+    We will use four "directions" for traversing it and a stack to store found pieces of land. 
+    We will traverse adjacent cells in search for next "land piece".
+    Each such adjacent "land piece" we will put in a stack.
+    We will process items in the stack until it's empty.
+    Then we will increase the number of islands.
+
+The time complexity will be O(N) since we iterate over each cell in matrix and track visited nodes in dictionary when processing.
+The space complexity will be also O(N).
+*/
+
+using System.Diagnostics;
+
+Tests.Test1();
+Tests.Test2();
+
+public static class Tests
+{
+    public static void Test1()
+    {
+        Console.WriteLine(nameof(Test1));
+        var sol = new Solution();
+
+        char [][] a = {
+            new char[] {'1', '1', '0', '0', '0'} ,
+            new char[] {'1', '1', '0', '0', '0'} ,
+            new char[] {'0', '0', '1', '0', '0'} ,
+            new char[] {'0', '0', '0', '1', '1'}
+        };
+
+        char res = sol.NumIslands(a);
+        Debug.Assert(res==3);
+        Console.WriteLine($"Number of islands: {(int)res}");
+    }
+
+    public static void Test2()
+    {
+        Console.WriteLine(nameof(Test2));
+        var sol = new Solution();
+
+        char [][] a = {
+            new char[] {'1', '0', '1', '1', '1'} ,
+            new char[] {'1', '0', '1', '0', '1'} ,
+            new char[] {'1', '1', '1', '0', '1'} 
+        };
+
+        char res = sol.NumIslands(a);
+        Debug.Assert(res==1);
+        Console.WriteLine($"Number of islands: {(int)res}");
+    }
+}
+
+
+public class Solution {
+    public char NumIslands(char[][] grid) {
```

Comment:
Well described! Although, the solution itself feels complex. Who would have guessed that u will need 4 loops :fearful: